### PR TITLE
Ingela/ssl/downgrade/gh 5009/otp 17517

### PR DESCRIPTION
**Pull request by IngelaAndin**
Solve at least some issues discovered due to  #5009

### CODE REVIEW

File: lib/ssl/src/ssl_connection.hrl [hunk 1]
```diff
@@ -93,7 +93,8 @@
 -record(connection_env, { 
                           user_application      :: {Monitor::reference(), User::pid()},
                           downgrade,
```
Comment (KennethL) [inline lib/ssl/src/ssl_connection.hrl:95]: Why not have a type spec for the downgrade field as well. Should be `{Pid::pid(),From:pid()}` i think. The question is what to name the first element with a better name than Pid.

**Review by u3s — COMMENTED** [review]
I've left some minor comments inline.

File: lib/ssl/src/ssl_gen_statem.erl [hunk 2]
```diff
@@ -1051,13 +1046,13 @@ maybe_invalidate_session({false, first}, server = Role, Host, Port, Session) ->
...
-    %% when ssl:close/1 returns unless it is a downgrade where
-    %% we want to guarantee that close alert is received before
-    %% returning. In both cases terminate has been run manually
-    %% before run by gen_statem which will end up here
+terminate({shutdown, downgrade}, downgrade, State) ->
+    %% Socket shall not be closed as it sgould be returned to user
+    handle_trusted_certs_db(State);
+terminate(_, _, #state{connection_env = #connection_env{socket_terminated = true}}) ->
+    %% Happens when user closes the connection using ssl:close/1 e.i. terminate
+    %% is called explicitly beforhand to to guarantee that Transport:close has been called
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl_gen_statem.erl]: fixed

File: lib/ssl/src/ssl_gen_statem.erl [hunk 3]
```diff
@@ -1051,13 +1046,13 @@ maybe_invalidate_session({false, first}, server = Role, Host, Port, Session) ->
...
 
-terminate(_, _, #state{connection_env = #connection_env{terminated = true}}) ->
-    %% Happens when user closes the connection using ssl:close/1
-    %% we want to guarantee that Transport:close has been called
-    %% when ssl:close/1 returns unless it is a downgrade where
-    %% we want to guarantee that close alert is received before
-    %% returning. In both cases terminate has been run manually
-    %% before run by gen_statem which will end up here
+terminate({shutdown, downgrade}, downgrade, State) ->
+    %% Socket shall not be closed as it sgould be returned to user
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl_gen_statem.erl]: fixed

File: lib/ssl/src/ssl_connection.hrl (hunk 1 above)
Comment (IngelaAndin) [reply lib/ssl/src/ssl_connection.hrl]: I made it {NewController::pid(), Caller::gen_statem:from()}

File: lib/ssl/src/ssl_connection.hrl [hunk 4]
```diff
@@ -92,8 +92,9 @@
 
 -record(connection_env, { 
                           user_application      :: {Monitor::reference(), User::pid()},
-                          downgrade,
-                          terminated = false                          ::boolean() | closed,  
+                          downgrade             :: {NewController::pid(), Caller::gen_statem:from()} | 'undefined',
```
Comment (KennethL) [inline lib/ssl/src/ssl_connection.hrl:95]: If the element in the tuple are called NewController and Caller in the type spec why are the variable names not named the same? Now when matching the variables are named Pid and From but why not NewController and Caller instead?
Just a thought.

File: lib/ssl/src/ssl_gen_statem.erl [hunk 5]
```diff
@@ -596,16 +596,11 @@ connection({call, From}, negotiated_protocol,
...
-                                                                downgrade = {Pid, From}},
-                                        protocol_specific = PS#{active_n_toggle => true,
-                                                                active_n => 1}
-                                       },
-     [{next_event, internal, ?ALERT_REC(?WARNING, ?CLOSE_NOTIFY)}]};
-connection({call, From},
+           #state{connection_env = #connection_env{socket_tls_closed = true} = CEnv} = State) ->
+    %%Close Alert already received from peer, "replay" in downgrade state.
+    {next_state, downgrade, State#state{connection_env = CEnv#connection_env{downgrade = {Pid, From}}},
+                                        [{next_event, internal, ?ALERT_REC(?WARNING, ?CLOSE_NOTIFY)}]};
```
Comment (u3s) [inline lib/ssl/src/ssl_gen_statem.erl:603]: Should the tls_sender:downgrade/2 call go here as well or not?

File: lib/ssl/src/ssl_connection.hrl (hunk 4 above)
Comment (IngelaAndin) [reply lib/ssl/src/ssl_connection.hrl]: Well, to make it consistent I changed it again to  NewController and From as From is very commonly used to refer to the caller when using the gen_* behaviours.  And changed the code accordingly!

File: lib/ssl/src/ssl_gen_statem.erl (hunk 5 above)
Comment (IngelaAndin) [reply lib/ssl/src/ssl_gen_statem.erl]: Humm ... good question I actually think it should! I made a new version.

**Comment by IngelaAndin** [comment]
This is merged now as it is an improvement of downgrade handling, although I am not sure if it completely solves the reported issue. We have a known race condition that we are working on independently of this that involves the downgrade scenario. Merging this will facilitate the continued work on that.

**Pull request merged**